### PR TITLE
[MIDDLEWARE] isAdmin -OT102-52

### PR DESCRIPTION
**Pull request by maxicuyo94**
HOW: admin user
I WANT: restrict access to certain content
TO: keep data safe

Acceptance criteria: You must verify the roleId field, and allow to continue if the user has a role corresponding to administrator

I use jwt and attached examples to use it and to call it
image

### CODE REVIEW

File: config/config.js [hunk 1]
```diff
@@ -8,6 +8,7 @@ module.exports = {
     host: process.env.DB_HOST,
     port: process.env.DB_PORT,
     dialect: 'mysql',
+    secret: process.env.JWT_SECRET_KEY,
```
Comment (dvelezroman) [inline config/config.js:11]: good

File: middlewares/isAdmin.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+// midleware to check if the user.roleId user is admin
+const jwt = require('jsonwebtoken')
+const config = require('../config/config').development
+
+const isAdmin = (req, res, next) => {
+  const token = req.headers['x-access-token']
+  if (!token) {
+    return res.status(401).send({ auth: false, message: 'No token provided.' })
```
Comment (dvelezroman) [inline middlewares/isAdmin.js]: apply the error handler here...

File: middlewares/isAdmin.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  jwt.verify(token, config.secret, (err, decoded) => {
+    if (err) {
+      return res
+        .status(500)
```
Comment (dvelezroman) [inline middlewares/isAdmin.js]: error handler here as well

File: middlewares/isAdmin.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        .send({ auth: false, message: 'Failed to authenticate token.' })
```
Comment (dvelezroman) [inline middlewares/isAdmin.js]: use .json

File: middlewares/isAdmin.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    if (decoded.roleId === 1) {
```
Comment (dvelezroman) [inline middlewares/isAdmin.js]: what if Admin has another roleId ???  we should assure to compare with the role name.

File: middlewares/isAdmin.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+      next()
+    } else {
+      return res
+        .status(403)
```
Comment (dvelezroman) [inline middlewares/isAdmin.js]: same as above.. errorhandler

File: middlewares/isAdmin.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        .send({ auth: false, message: 'You are not an admin' })
```
Comment (dvelezroman) [inline middlewares/isAdmin.js]: .json

File: middlewares/isAdmin.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    return null
+  })
+  return null
+}
+module.exports = isAdmin
+
+// examples
```
Comment (dvelezroman) [inline middlewares/isAdmin.js]: instead of this commented code... you could do : 
a document of this is README.md

File: routes/ping.js [hunk 9]
```diff
@@ -2,7 +2,9 @@ const express = require('express')
 
 const router = new express.Router()
 const { pong } = require('../controllers/ping')
+const isAdmin = require('../middlewares/isAdmin')
 
 router.get('/', pong)
+router.get('/admin', isAdmin, pong)
```
Comment (dvelezroman) [inline routes/ping.js]: this route is only for testing purposes.. to see if the sever is online... please do not use middleware for admin access here...

**Review by dvelezroman — CHANGES_REQUESTED** [review]
Good PR .... few comments.

**Comment by maxicuyo94** [comment]
fixed requested

File: middlewares/isAdmin.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
+// midleware to check if the user.roleId user is admin
+const jwt = require('jsonwebtoken')
+const config = require('../config/config').development
+const ErrorObject = require('../helpers/error')
+
+const isAdmin = (req, res, next) => {
+  const token = req.headers['x-access-token']
+  if (!token) {
+    next(new ErrorObject('No token provided.', 401))
```
Comment (JonatanGarbuyo) [inline middlewares/isAdmin.js]: si pasas el error con next no se devuelve correctamente
hay que devolver con res.json()

fijate en el  middleware de validateRequest

File: middlewares/isAdmin.js [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  }
+  jwt.verify(token, config.secret, (err, decoded) => {
+    if (err) {
+      next(new ErrorObject('Failed to authenticate token.', 401))
+    }
+    if (decoded.role === 'Admin') {
+      next()
+    } else {
+      next(new ErrorObject('You are not an admin', 403))
```
Comment (JonatanGarbuyo) [inline middlewares/isAdmin.js]: devolver json() idem anterior.
ademas, solo debería devolver 'Forbidden' sin develar el funcionamiento interno.

File: middlewares/isAdmin.js [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+const ErrorObject = require('../helpers/error')
+
+const isAdmin = (req, res, next) => {
+  const token = req.headers['x-access-token']
+  if (!token) {
+    next(new ErrorObject('No token provided.', 401))
+  }
+  jwt.verify(token, config.secret, (err, decoded) => {
+    if (err) {
+      next(new ErrorObject('Failed to authenticate token.', 401))
```
Comment (JonatanGarbuyo) [inline middlewares/isAdmin.js]: devolver con res.json()

**Review by JonatanGarbuyo — COMMENTED** [review]
esta muy bien. te dejo algunos comentarios

File: middlewares/isAdmin.js (hunk 10 above)
Comment (dvelezroman) [reply middlewares/isAdmin.js]: con las modificaciones al error handler .. esto cambia... fijarse en la rama develop y el controller de user .

**Review by dvelezroman — CHANGES_REQUESTED** [review]

File: middlewares/isAdmin.js [hunk 13]
```diff
@@ -0,0 +1,24 @@
+// midleware to check if the user.roleId user is admin
+const jwt = require('jsonwebtoken')
+const config = require('../config/config').development
+const ErrorObject = require('../helpers/error')
+
+const isAdmin = (req, res, next) => {
+  const token = req.headers['x-access-token']
+  if (!token) {
+    next(new ErrorObject('No token provided.', 401))
+  }
```
Comment (JonatanGarbuyo) [inline middlewares/isAdmin.js]: te faltó usar el  createHttpError.  por ej:

const httpError = createHttpError( 401,  `[ No token provided. ] - [ ${req.baseUrl} - ${req.method}]`,  )
next(httpError)

sino'  tira mal el statusCode

**Review by JonatanGarbuyo — APPROVED** [review]

**Review by dvelezroman — APPROVED** [review]

**Review by gabodmc — APPROVED** [review]

**Review by gabyMangoni — APPROVED** [review]
Approve

**Pull request merged**